nifi-commons/nifi-record/src/test/java/org/apache/nifi/serialization/record/TestDataTypeUtils.java

```diff
@@ -324,4 +332,199 @@ public void testConvertDataTypeBigint() {
         }
         assertNotNull(e);
     }
+    @Test
+    public void testChooseDataTypeWhenInt_vs_INT_FLOAT_ThenShouldReturnINT() {
+        // GIVEN
+        List<DataType> dataTypes = Arrays.asList(
+                RecordFieldType.INT.getDataType(),
+                RecordFieldType.FLOAT.getDataType()
+        );
+
+        Object value = 1;
+        DataType expected = RecordFieldType.INT.getDataType();
+
+        // WHEN
+        // THEN
+        runWithNormalAndReverseInput(this::testChooseDataType, input(dataTypes, value), expected);
+    }
+
+    @Test
+    public void testChooseDataTypeWhenFloat_vs_INT_FLOAT_ThenShouldReturnFLOAT() {
+        // GIVEN
+        List<DataType> dataTypes = Arrays.asList(
+                RecordFieldType.INT.getDataType(),
+                RecordFieldType.FLOAT.getDataType()
+        );
+
+        Object value = 1.5f;
+        DataType expected = RecordFieldType.FLOAT.getDataType();
+
+        // WHEN
+        // THEN
+        runWithNormalAndReverseInput(this::testChooseDataType, input(dataTypes, value), expected);
+    }
+
+    @Test
+    public void testChooseDataTypeWhenHasChoiceThenShouldReturnSingleMatchingFromChoice() {
+        // GIVEN
+        List<DataType> dataTypes = Arrays.asList(
+                RecordFieldType.INT.getDataType(),
+                RecordFieldType.DOUBLE.getDataType(),
+                RecordFieldType.CHOICE.getChoiceDataType(
+                        RecordFieldType.FLOAT.getDataType(),
+                        RecordFieldType.STRING.getDataType()
+                )
+        );
+
+        Object value = 1.5f;
+        DataType expected = RecordFieldType.FLOAT.getDataType();
+
+        // WHEN
+        // THEN
+        runWithNormalAndReverseInput(this::testChooseDataType, input(dataTypes, value), expected);
+    }
+
+    private <E> void runWithNormalAndReverseInput(BiFunction<Input, E, ?> test, Input input, E expected) {
```

Comment:
It looks like the first parameter is always the same

Reply:
Yeah, in the end this `BiFunction` became unnecessary.